### PR TITLE
Epic Script

### PR DESCRIPTION
**Pull request by sipatra11**
Added Support for metadata handling. 

Case 1: if file changes export all
Case 2: Filename same but lastwritetime changed write all
Case 3: Filename same and lastwrite time same, check continuation token, skip lines and then start to push

### CODE REVIEW

**Review by vakataka — CHANGES_REQUESTED** [review]
Can we update the other script itself ?

File: epic-sample-script.ps1 [hunk 1]
```diff
@@ -0,0 +1,294 @@
+param
```
Comment (suds05) [inline epic-sample-script.ps1:1]: Would be nice to have some comment at top to indicate purpose of the script

File: epic-sample-script.ps1 [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+    $ErrorRecord = [System.Management.Automation.ErrorRecord]::new(
+        $Exception,
+        "errorID",
+        [System.Management.Automation.ErrorCategory]::NotSpecified,
+        $TargetObject
+    )
+    $PSCmdlet.WriteError($ErrorRecord)
+}
+
+function UploadData ($access_token, $FileName) {
```
Comment (suds05) [inline epic-sample-script.ps1:105]: Isn't the top level function supposed to be in a Verb-Noun format (Upload-HealthEvents or something like that)...?

Comment (suds05) [inline epic-sample-script.ps1:105]: Good to annotate top level function with usage information. See https://github.com/microsoft/CSS-Exchange/blob/main/Security/src/Test-ProxyLogon.ps1

File: epic-sample-script.ps1 [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+    $stream.Position = 0
+    $filemetaData = Get-FileHash -InputStream $stream | Select-Object Hash
+    $stream.Dispose()
+    $writer.Dispose()
+    return $filemetaData
+}
+
+function ChunkAndSend($FileName, $linesperFile) {
+    $DirName = [System.IO.Path]::GetDirectoryName($FileName)
+    $TmpFileName = "\tmp"
```
Comment (suds05) [inline epic-sample-script.ps1:224]: Shouldn't temp data be created under $env:TEMP. Would we have permission to create files in the path specified.

File: epic-sample-script.ps1 [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+            # Skip no of rows already written as per metadata
+            while ($activeLineCount -lt $metaData.NoOfRowsWritten -and $reader.EndOfStream -ne $true) {
+                $reader.ReadLine() | Out-Null
+                $activeLineCount++
+            }
+            
+            while ($reader.EndOfStream -ne $true) {              
+                $linecount = 0
+                $NewFile = "{0}{1}{2}{3}" -f ($DirName, $TmpFileName, $filecount.ToString("0000"), $ext)
+                "Creating file $NewFile"
```
Comment (suds05) [inline epic-sample-script.ps1:250]: Shouldn't this be a Write-Output or a Write-Verbose.

File: sample_script.ps1 [hunk 5]
```diff
@@ -1,3 +1,12 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+# 
+# Sample Script to Push Data to M365 Compliance Connector 
+# Examples
+#
+# Push data to M365 Compliance Connector:
+# .\sample_audit_script_new.ps1 -tenantId <Guid> -appId <App Id> -appSecret <App Secret> -jobId <Job id GUID> -filePath <File Path> -Verbose
```
Comment (sasanche) [inline sample_script.ps1:8]: nit: instead of sample_audit_script_new.ps1 use actual script name.

**Comment by sipatra11** [comment]
> Can we update the other script itself ?

@vakataka , have updated the other script. Can you please review the changes again.

**Comment by sipatra11** [comment]
> Can we update the other script itself ?

File: sample_script.ps1 [hunk 6]
```diff
@@ -143,12 +171,149 @@ function UploadData ($access_token) {
...
+        Checks for existing file metadata. Metadata has attributes corresponding to the last row for the file that was successfully sent. 
+        Use the metadata for checkpointing and then compute start point. 
+        Chunks file into specified size and send
+    .PARAMETER FileName
+        The File to be used for consumption and pushing Data
+    .PARAMETER LinesPerFile
+        Defines the Chunk Size
+#>
+function Send-ChunkedData($FileName, $linesperFile) {
+    #$DirName = [System.IO.Path]::GetDirectoryName($FileName)
```
Comment (sasanche) [inline sample_script.ps1:243]: Can be removed?

File: sample_script.ps1 [hunk 7]
```diff
@@ -143,12 +171,149 @@ function UploadData ($access_token) {
...
+
+    if ( !(Test-Path $TmpDirName -PathType Container)) {
+        New-Item -ItemType directory -Path $TmpDirName
+    }
+
+    $TmpFileName = "\tmp"
+    $ext = ".txt"
+    #$linesperFile = 10#100k
```
Comment (sasanche) [inline sample_script.ps1:251]: Can be removed?

File: sample_script.ps1 [hunk 8]
```diff
@@ -143,12 +171,149 @@ function UploadData ($access_token) {
...
+    $filecount = 1
+    $reader = $null
+  
+    try {
+        $reader = [io.file]::OpenText($Filename)
+        # Handle Obselete Metadata
+        HandleObseleteMetadata
```
Comment (sasanche) [inline sample_script.ps1:258]: nit: spell correction Obsolete

Comment (sasanche) [reply sample_script.ps1:258]: handleObsoleteMetadata can be called within GetOrCreateMetadata

File: sample_script.ps1 [hunk 9]
```diff
@@ -143,12 +171,149 @@ function UploadData ($access_token) {
...
+        try {        
+            $header = $reader.ReadLine();
+            $activeLineCount = 0
+
+            # Skip no of rows already written as per metadata
+            while ($activeLineCount -lt $metaData.NoOfRowsWritten -and $reader.EndOfStream -ne $true) {
+                $reader.ReadLine() | Out-Null
+                $activeLineCount++
+            }
+
```
Comment (sasanche) [inline sample_script.ps1:271]: Can add a log that these many lines were skipped

File: sample_script.ps1 [hunk 10]
```diff
@@ -9,17 +18,28 @@ param
...
 $eventApiEndpoint = "api/signals"
 
-function GetAccessToken () {
+$serviceName = "PushConnector"
+$TmpDirName = $env:TEMP + "ctr###D"
+
+class FileMetdata {
+    [string]$FileHash
+    [string]$NoOfRowsWritten
+    [string]$Service
```
Comment (sasanche) [inline sample_script.ps1:39]: Can add time stamp of last update time.

File: sample_script.ps1 [hunk 11]
```diff
@@ -143,12 +171,149 @@ function UploadData ($access_token) {
...
+                #"Closing file"
+                $writer.Dispose();
+
+                Write-Verbose "Created file with $linecount records"
+                RetryCommand -ScriptBlock {
+                    param($fileName)
+                    $access_token = Get-AccessToken
+                    Push-Data($access_token, $fileName) $NewFile
+                }
+
```
Comment (sasanche) [inline sample_script.ps1:299]: We shall update metadata after every push

File: sample_script.ps1 [hunk 12]
```diff
@@ -1,3 +1,12 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+# 
+# Sample Script to Push Data to M365 Compliance Connector 
+# Examples
```
Comment (sasanche) [inline sample_script.ps1:5]: can add synopsis,description, parameter, examples.

File: sample_script.ps1 [hunk 13]
```diff
@@ -143,12 +171,149 @@ function UploadData ($access_token) {
...
+        Write-ErrorMessage("Error body : {0}" -f $result.Content.ReadAsStringAsync().Result)
     }
 }
 
-RetryCommand -ScriptBlock {
-    $access_token = GetAccessToken
-    UploadData($access_token)
+function GetOrCreateMetadata($FileName) {
+    $fileHash = ComputeHashForInputFile($FileName)
+    $metaDataFileName = $TmpDirName + "\." + $fileHash + "####mdata.txt"
```
Comment (sasanche) [inline sample_script.ps1:181]: Create a helper function for creating this file name and constant string for suffix, being used at multiple places.

File: sample_script.ps1 [hunk 14]
```diff
@@ -143,12 +171,149 @@ function UploadData ($access_token) {
...
+    $filemetaData.NoOfRowsWritten = $noOfRowsWritten
+    $metaDataFilePath = $TmpDirName + "\." + $fileHash + "####mdata.txt"
+    $filemetaData | ConvertTo-Json -Depth 100 | Out-File $metaDataFilePath
+}
+
+function HandleObseleteMetadata() {
+    # Delete metadata which are over 14 days old
+    $timeLimit = (Get-Date).AddDays(-14)
+    $filePath = $TmpDirName
+    Get-ChildItem -Path $filePath -Recurse -Force | Where-Object { !$_.PSIsContainer -and $_.LastWriteTime -lt $timeLimit } | Where-Object { $_.Name -match '^.+\####mdata.txt$' } | Remove-Item -Force
```
Comment (sasanche) [inline sample_script.ps1:213]: nit: multiline.

**Review by sasanche — COMMENTED** [review]
Added a few comments. Good to add metadata content detail in PR description.

**Review by vakataka — APPROVED** [review]

File: sample_script.ps1 [hunk 15]
```diff
@@ -1,3 +1,36 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+# 
+# Sample Script to Push Data to M365 Compliance Connector 
+# Examples
+#
+# Push data to M365 Compliance Connector:
+# .\sample_script.ps1 -tenantId <Guid> -appId <App Id> -appSecret <App Secret> -jobId <Job id GUID> -filePath <File Path> -Verbose
```
Comment (sasanche) [inline sample_script.ps1:8]: minor: this can be added as example in below synposis, description section

**Review by sasanche — APPROVED** [review]

**Pull request merged**